helpers/discordAuth.ts

```diff
@@ -0,0 +1,111 @@
+import { prisma } from '../prisma'
+import { URLSearchParams } from 'url'
+import { User } from '.prisma/client'
+
+const discordAPI = 'https://discordapp.com/api'
+const client_id = process.env.DISCORD_KEY
+const client_secret = process.env.DISCORD_SECRET
+
+type AccessTokenResponse = {
+  access_token: string
+  token_type: string
+  expires_in: number
+  refresh_token: string
+  scope: string
+}
+
+type UserInfoResponse = {
+  id: string
+  username: string
+  avatar: string
+  discriminator: string
+  public_flags: number
+  locale: string
+  mfa_enabled: boolean
+  email: string
+  verified: boolean
+}
+
+type DiscordUserInfo = {
+  discordUserId: string
+  discordUsername: string
+  discordAvatarUrl: string
+  discordRefreshToken: string
+}
+
+export const getTokenFromAuthCode = (
+  code: string
+): Promise<AccessTokenResponse> => {
+  return fetch(`${discordAPI}/oauth2/token`, {
+    method: 'POST',
+    headers: {
+      'content-type': 'application/x-www-form-urlencoded;charset=utf-8'
+    },
+    body: new URLSearchParams({
+      grant_type: 'authorization_code',
+      client_id,
+      client_secret,
+      code,
+      redirect_uri: 'https://c0d3.com/discord/redir',
+      scope: 'email guilds.join gdm.join identify'
+    })
+  }).then(r => r.json())
+}
+
+export const getTokenFromRefreshToken = (
+  refresh_token: string
+): Promise<AccessTokenResponse> => {
+  return fetch(`${discordAPI}/oauth2/token`, {
+    method: 'POST',
+    headers: {
+      'content-type': 'application/x-www-form-urlencoded;charset=utf-8'
+    },
+    body: new URLSearchParams({
+      client_id,
+      client_secret,
+      grant_type: 'refresh_token',
+      refresh_token
+    })
+  }).then(r => r.json())
+}
+
+export const getUserInfo = (accessToken: string): Promise<UserInfoResponse> => {
+  return fetch(`${discordAPI}/users/@me`, {
+    method: 'GET',
+    headers: {
+      'content-type': 'application/x-www-form-urlencoded;charset=utf-8',
+      Authorization: `Bearer ${accessToken}`
+    }
+  }).then(r => r.json())
+}
+
+export const updateUserRefreshToken = async (
```

Comment:
do we want to export this function? I was under the impression that only `getUserInfoFromRefreshToken` is required, and therefore only that function should be exported?

Reply:
true, i removed the two functions from exports. the others may be needed.

Reply:
oh I remembered... you must always call `updateUserRefreshToken` after getting the accessToken. This is because after getting user's accessToken, the refresh token becomes invalid and you need to get a new one. so this function is a helper function that shouldn't be exported.

Reply:
I see now. I have a case during login where I just need to check if refreshToken exists and is valid, and no userInfo needs to be fetched. Therefore I just store the new refreshToken in that case, using the functions getTokenFromRefreshToken and updateUserRefreshToken

![121316637-3292ae00-c912-11eb-93d9-12d7dc675ab9](https://user-images.githubusercontent.com/16906103/123053336-74d8e680-d40c-11eb-9d76-ed92fbf4f6a6.png)

Reply:
yeah so from what you are saying, only `getUserInfoFromRefreshToken` needs to be exported right?